File: scanner.py
from token import Token, TokenType

class Scanner:
    def __init__(self, source: str):
        self.source = source
        self.tokens = []
        self.start = 0
        self.current = 0
        self.line = 1

    def scan_tokens(self):
        while not self.is_at_end():
            self.start = self.current
            self.scan_token()
        self.tokens.append(Token(TokenType.EOF, "", None, self.line))
        return self.tokens

    def scan_token(self):
        c = self.advance()

        # --- single-character tokens ---
        if c == '(': self.add_token(TokenType.LEFT_PAREN)
        elif c == ')': self.add_token(TokenType.RIGHT_PAREN)
        elif c == '{': self.add_token(TokenType.LEFT_BRACE)
        elif c == '}': self.add_token(TokenType.RIGHT_BRACE)
        elif c == ',': self.add_token(TokenType.COMMA)
        elif c == '.': self.add_token(TokenType.DOT)
        elif c == '-': self.add_token(TokenType.MINUS)
        elif c == '+': self.add_token(TokenType.PLUS)
        elif c == ';': self.add_token(TokenType.SEMICOLON)
        elif c == '*': self.add_token(TokenType.STAR)
        elif c == '/': self.add_token(TokenType.SLASH)
        elif c == '=': self.add_token(TokenType.EQUAL_EQUAL if self.match('=') else TokenType.EQUAL)
        elif c == '!': self.add_token(TokenType.BANG_EQUAL if self.match('=') else TokenType.BANG)
        elif c == '<': self.add_token(TokenType.LESS_EQUAL if self.match('=') else TokenType.LESS)
        elif c == '>': self.add_token(TokenType.GREATER_EQUAL if self.match('=') else TokenType.GREATER)

        # --- string literal ---
        elif c == '"':
            self.string()

        # --- number literal ---
        elif c.isdigit():
            self.number()

        # ignore whitespace
        elif c in [' ', '\r', '\t']:
            pass
        elif c == '\n':
            self.line += 1
        else:
            print(f"[Line {self.line}] Unexpected character: {c}")

    # helpers
    def advance(self):
        self.current += 1
        return self.source[self.current - 1]

    def match(self, expected):
        if self.is_at_end(): return False
        if self.source[self.current] != expected: return False
        self.current += 1
        return True

    def peek(self):
        if self.is_at_end(): return '\0'
        return self.source[self.current]

    def peek_next(self):
        if self.current + 1 >= len(self.source): return '\0'
        return self.source[self.current + 1]

    def is_at_end(self):
        return self.current >= len(self.source)

    def add_token(self, type_, literal=None):
        text = self.source[self.start:self.current]
        self.tokens.append(Token(type_, text, literal, self.line))

    # string handling
    def string(self):
        while self.peek() != '"' and not self.is_at_end():
            if self.peek() == '\n': self.line += 1
            self.advance()

        if self.is_at_end():
            print(f"[Line {self.line}] Error: Unterminated string.")
            return

        # closing quote
        self.advance()
        value = self.source[self.start + 1:self.current - 1]
        self.add_token(TokenType.STRING, value)

    # number handling
    def number(self):
        while self.peek().isdigit():
            self.advance()

        if self.peek() == '.' and self.peek_next().isdigit():
            self.advance()
            while self.peek().isdigit():
                self.advance()

        value = float(self.source[self.start:self.current])
        self.add_token(TokenType.NUMBER, value)

File: Scanner.kt
class Scanner(private val source: String) {
    private val tokens = mutableListOf<Token>()
    private var start = 0
    private var current = 0
    private var line = 1

    // all keyword keys are lowercase (case-insensitive)
    private val keywords = mapOf(
        // expressions
        "mix" to TokenType.MIX,
        "take away" to TokenType.TAKE_AWAY,
        "multiply" to TokenType.MULTIPLY,
        "divide" to TokenType.DIVIDE,
        "flip" to TokenType.FLIP,
        "check if" to TokenType.CHECK_IF,

        // statements
        "remember" to TokenType.REMEMBER,
        "set" to TokenType.SET,
        "print" to TokenType.PRINT,
        "as" to TokenType.AS,
        "to" to TokenType.TO,

//        // natural language sugar
//        "the" to TokenType.THE,
//        "value" to TokenType.VALUE,
//        "when" to TokenType.WHEN,
//        "you" to TokenType.YOU,

        // cookbook-style control flow (your choice)
        "when" to TokenType.IF,          // cookbook 'when' -> if
        "otherwise" to TokenType.ELSE,   // cookbook 'otherwise' -> else
        "repeat" to TokenType.WHILE,     // 'repeat' -> while
        "for" to TokenType.FOR,          // 'for' -> for
        "recipe" to TokenType.FUN,       // 'recipe' -> function
        "serve" to TokenType.RETURN,     // 'serve' -> return

        // connectors
        "and" to TokenType.AND,
        "from" to TokenType.FROM,
        "with" to TokenType.WITH,
        "or" to TokenType.OR,

        // boolean
        "true" to TokenType.TRUE,
        "false" to TokenType.FALSE,

    )

    fun scanTokens(): List<Token> {
        while (!isAtEnd()) {
            start = current
            scanToken()
        }
        tokens.add(Token(TokenType.EOF, "", null, line))
        return tokens
    }

    private fun scanToken() {
        val c = advance()
        when {
            c.isWhitespace() -> {
                if (c == '\n') line++
            }
            c.isDigit() -> number()
            c == '"' -> string()
            c.isLetter() -> identifier()

            c == '{' -> addToken(TokenType.LEFT_BRACE)
            c == '}' -> addToken(TokenType.RIGHT_BRACE)
            c == '(' -> addToken(TokenType.LEFT_PAREN)
            c == ')' -> addToken(TokenType.RIGHT_PAREN)
            c == ';' -> addToken(TokenType.SEMICOLON)
            c == ',' -> addToken(TokenType.COMMA)

            c == '>' -> addToken(TokenType.GREATER)
            c == '<' -> addToken(TokenType.LESS)
            c == '=' && match('=') -> addToken(TokenType.EQUAL_EQUAL)

            else -> error("Unexpected character '$c'")
        }
    }

    private fun identifier() {
        while (peek().isLetterOrDigit() || peek() == '_') advance()

        val original = source.substring(start, current)
        var lower = original.lowercase()

        // handle two-word keywords like "take away", "check if"
        if (peek().isWhitespace()) {
            val saved = current
            skipWhitespace()
            val nextWordStart = current

            while (peek().isLetter()) advance()
            val nextWord = source.substring(nextWordStart, current).lowercase()

            val combined = "$lower $nextWord"
            if (keywords.containsKey(combined)) {
                lower = combined
            } else {
                current = saved
            }
        }


        val type = keywords[lower]
        if (type != null) addToken(type)
        else addToken(TokenType.IDENTIFIER, original) // case-sensitive
    }

    private fun skipWhitespace() {
        while (peek().isWhitespace() && peek() != '\n') advance()
    }

    private fun number() {
        while (peek().isDigit()) advance()
        if (peek() == '.' && peekNext().isDigit()) {
            advance()
            while (peek().isDigit()) advance()
        }
        val value = source.substring(start, current).toDouble()
        addToken(TokenType.NUMBER, value)
    }

    private fun string() {
        while (peek() != '"' && !isAtEnd()) {
            if (peek() == '\n') line++
            advance()
        }
        if (isAtEnd()) {
            error("Unterminated string.")
            return
        }
        advance()
        val value = source.substring(start + 1, current - 1)
        addToken(TokenType.STRING, value)
    }

    private fun advance(): Char {
        current++
        return source[current - 1]
    }

    private fun match(expected: Char): Boolean {
        if (isAtEnd()) return false
        if (source[current] != expected) return false
        current++
        return true
    }

    private fun peek(): Char =
        if (isAtEnd()) '\u0000' else source[current]

    private fun peekNext(): Char =
        if (current + 1 >= source.length) '\u0000' else source[current + 1]

    private fun isAtEnd() = current >= source.length

    private fun addToken(type: TokenType, literal: Any? = null) {
        tokens.add(Token(type, source.substring(start, current), literal, line))
    }

    private fun error(message: String) {
        println("[Line $line] Error: $message")
    }
}
